lib/absinthe/ecto.ex

```diff
@@ -115,47 +115,37 @@ defmodule Absinthe.Ecto do
   end
   """
   def ecto_batch(repo, %model{} = parent, association, callback \\ &default_callback/1) do
-    case model.__schema__(:association, association) do
-      %Ecto.Association.BelongsTo{} = assoc ->
-        build_batch(:perform_belongs_to, repo, parent, assoc, callback)
-      %Ecto.Association.Has{cardinality: :many} = assoc ->
-        build_batch(:perform_has_many, repo, parent, assoc, callback)
-    end
-  end
+    assoc = model.__schema__(:association, association)
+
+    %{owner: owner,
+      owner_key: owner_key,
+      field: field} = assoc
 
-  defp build_batch(batch_fun, repo, parent, assoc, callback) do
-    id = Map.fetch!(parent, assoc.owner_key)
+    id = Map.fetch!(parent, owner_key)
 
-    meta = {repo, assoc.queryable, assoc.related_key, self()}
+    meta = {repo, owner, owner_key, field, self()}
 
-    batch({__MODULE__, batch_fun, meta}, id, fn results ->
+    batch({__MODULE__, :perform_batch, meta}, id, fn results ->
       results
-      |> Map.get(id, default_result(batch_fun))
+      |> Map.get(id, default_result(assoc))
       |> callback.()
     end)
   end
 
-  defp default_result(:perform_belongs_to), do: nil
-  defp default_result(:perform_has_many), do: []
+  defp default_result(%Ecto.Association.BelongsTo{}), do: nil
+  defp default_result(%Ecto.Association.Has{cardinality: :many}), do: []
+  defp default_result(%Ecto.Association.Has{}), do: nil
+  defp default_result(%Ecto.Association.ManyToMany{}), do: []
```

Comment:
Every one of those structs should have a `cardinality` field - it should be enough to match on those:

```elixir
defp default_result(%{cardinality: :one}), do: nil
defp default_result(%{cardinality: :many}), do: []
```